### ISSUE TITLE
pathlib strips trailing slash

### ISSUE THREAD

**Issue by 4980da3a-3107-46a8-b98d-461c0fb8b0e7** [python/cpython#65238]
BPO | [21039](https://bugs.python.org/issue21039)
--- | :---
Nosy | @akuchling, @pitrou, @serhiy-storchaka, @sigmavirus24, @hvenev
Files | <li>[pathlib.patch](https://bugs.python.org/file34586/pathlib.patch "Uploaded as text/plain at 2014-03-23.13:55:55 by @hvenev"): patch</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2015-04-14.16:54:25.933>
created_at = <Date 2014-03-23.13:55:56.401>
labels = ['type-bug', 'library']
title = 'pathlib strips trailing slash'
updated_at = <Date 2015-04-14.16:54:25.932>
user = 'https://github.com/hvenev'
```

bugs.python.org fields:
```python
activity = <Date 2015-04-14.16:54:25.932>
actor = 'pitrou'
assignee = 'none'
closed = True
closed_date = <Date 2015-04-14.16:54:25.933>
closer = 'pitrou'
components = ['Library (Lib)']
creation = <Date 2014-03-23.13:55:56.401>
creator = 'h.venev'
dependencies = []
files = ['34586']
hgrepos = []
issue_num = 21039
keywords = ['patch']
message_count = 15.0
messages = ['214581', '214594', '214595', '214596', '214597', '214598', '214599', '214600', '214601', '214604', '224889', '224942', '240930', '240947', '240948']
nosy_count = 7.0
nosy_names = ['akuchling', 'pitrou', 'BreamoreBoy', 'serhiy.storchaka', 'icordasc', 'h.venev', 'ischwabacher']
pr_nums = []
priority = 'normal'
resolution = 'wont fix'
stage = None
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue21039'
versions = ['Python 3.4', 'Python 3.5']
```

</p></details>

### Blocks issues

- #64978
- #73991
- #102783
- #106747

### Linked PRs
* gh-103595
* gh-112363
* gh-113248

**Comment by barneygale** [comment]
I'm ~7 years late here, but I think this might be worth revisiting.

Pathlib's path normalization is conservative and tries hard not to change the meaning of paths. Some users complain that collapsing `..` segments is unfriendly, or that treating `//` and `/` as different roots is unnecessary, but it's the safe thing to do.

This is one of only two cases (that I'm aware of) where pathlib's normalization can change the meaning of a path (the other is #80486).

> For example, what should be the last component of the path? The component just before the ending slash, or the empty string? What if slice off the last part? What is the name, stem, suffix? etc.

The name, stem and suffix should probably be empty to align with `PurePath("/")` IMO.

**Comment by barneygale** [comment]
I will work on this once #102476 lands. Turns out it's necessary for the optimization I'm going for in #101560.

**Comment by shiftagain** [comment]
For a script I'm currently writing, I have to append a directory path to sys.path in order to make some Python imports work properly, and it seems that sys.path interprets a path differently depending on if it has a trailing slash or not.

I'm sure there are many other examples of code in the Python ecosystem requiring a trailing slash or behaving differently without one, so arguments about how pathlib is "Only for Python internals" or "Just represents a filesystem path" don't make sense to me.

I would be content with just adding a `resolve_with_trailing_slash` function in addition to the `resolve` function. I think it would make sense to change the default behavior to have a trailing slash for folders (so that you have the additional information in the raw path that it is a folder), but breaking changes are unfortunate.

**Comment by gvanrossum** [comment]
(Cross-posted to Discourse: https://discuss.python.org/t/pathlib-preserve-trailing-slash/33389/9)

I'm sorry Antoine chose to let `pathlib` deviate from the `os.path` module's behavior for this. I recall thinking long and hard about edge cases like this and carefully implementing what I thought was best. I haven't read that POSIX standard but I presume I was influenced by actual behavior of various UNIX utilities.

But I agree that changing this will break plenty of user code relying on the current behavior, and we can't have that. I also don't think that some kind of deprecation path makes sense here. The best we can do is have some way to indicate the preferred behavior when a path is created, and have that inherited by operations that return new paths.

I'm not sure what form the user preference should take -- I'd say it shouldn't be global, so it could take the form of either an alternative class, an alternative constructor, or a flag keyword argument to path-constructing operations.

Even so, there could be problems -- suppose we have a library that accepts `Path` arguments and expects them to behave the old way, and a user constructs paths using the alternate constructor and passes those in. It might be quite a while before the user ends up passing a path that causes the library to crash or misbehave.

So maybe an alternative approach could be not to have the behavior be indicated by some property of the `Path` instance but by using different attributes. So maybe e.g. `Path("foo/").name` would return `"foo"` but `Path("foo/").alt_name` would return `""`. This would avoid the scenario I described just above.

Now we just have to decide on names for the attributes that could have this alternate behavior (are there others besides `.parent` and `.name`?). And probably the implementation will have to keep track of the trailing slash somehow.

**Comment by barneygale** [comment]
> So maybe an alternative approach could be not to have the behavior be indicated by some property of the `Path` instance but by using different attributes. So maybe e.g. `Path("foo/").name` would return `"foo"` but `Path("foo/").alt_name` would return `""`. This would avoid the scenario I described just above.
> 
> Now we just have to decide on names for the attributes that could have this alternate behavior (are there others besides `.parent` and `.name`?). And probably the implementation will have to keep track of the trailing slash somehow.

I've been working on a patch along these lines. The idea is to ignore any trailing slash whenever we split a path into (dirname, basename). So:

```python
>>> from pathlib import PurePosixPath
>>> p = PurePosixPath('/home/barney/')
>>> p.parent
PurePosixPath('/home')
>>> list(p.parents)
[PurePosixPath('/home'), PurePosixPath('/')]
>>> p.name
'barney'
>>> p.with_name('fred')
PurePosixPath('/home/fred/')
```

This also applies to `[with_]stem`, `[with_]suffix`, `suffixes`, `[is_]relative_to`.

Users wouldn't be able to call `path.parent` to remove a trailing slash, so we'd need to add something like `PurePath.[with_]trailer` I think.

**Comment by barneygale** [comment]
Re-resolving as "won't fix" - the present behaviour is desirable for many users and it's too dangerous to change now. The crux of the issue is:

```python
>>> str(PurePath('a/')) == 'a'
True
>>> PurePath('a/') == PurePath('a')
True
```

Fixing this bug entails changing the first result, which entails changing the second too. For some this is clearly a bugfix, but for many others it comes close to defeating the purpose of pathlib.

I don't think an initialiser argument would help: users handed a `Path` object would still need to cope with the possible equality and string representation change.

Much more discussion here: https://discuss.python.org/t/pathlib-preserve-trailing-slash/33389/
